palace/utils/iodata.cpp

```diff
@@ -16,14 +20,131 @@
+                   R"(|(\/\*([^*]|[\r\n]|(\*+([^*\/]|[\r\n])))*\*+\/))"
+                   R"(|(\/\/.*))");
+    file = std::regex_replace(file, rgx, "$1");
+  }
+
+  // Also strip whitespace.
+  {
+    std::regex rgx(R"((([\"'])(?:(?=(\\?))\3.)*?\2))"
+                   R"(|(\s+))");
+    file = std::regex_replace(file, rgx, "$1");
+  }
+
+  // Also strip erroneous trailing commas.
+  {
+    std::regex rgx(R"((([\"'])(?:(?=(\\?))\3.)*?\2))"
+                   R"(|,+(?=\s*?[\}\]]))");
+    file = std::regex_replace(file, rgx, "$1");
+  }
+
+  // Perform integer range expansion for arrays ([a - b, c] = [a-b,c] =
+  // [a,a+1,...,b-1,b,c]). The whole file is now one line and arrays have no spaces after
+  // whitespace stripping.
+  std::stringstream output;
+  auto RangeExpand = [](std::string_view str) -> std::string
+  {
+    // Handle the given string which is only numeric with possible hyphens.
+    int num;
+    auto [ptr, ec] = std::from_chars(str.data(), str.data() + str.length(), num);
+    MFEM_VERIFY(ptr != str.data(), "Invalid integer conversion in range expansion!");
+    if (ptr == str.data() + str.length())
+    {
+      return std::string(str);
+    }
+    // Range specified, expand the bounds.
+    int num2;
+    auto [ptr2, ec2] = std::from_chars(ptr + 1, str.data() + str.length(), num2);
+    MFEM_VERIFY(ptr2 != ptr + 1, "Invalid integer conversion in range expansion!");
+    std::string rng;
+    while (num < num2)
+    {
+      rng += std::to_string(num++) + ",";
+    }
+    rng += std::to_string(num);
+    return rng;
+  };
+  {
+    const std::string range_vals = "-0123456789,";
+    auto start = file.begin();
+    bool inside = false;
+    for (auto it = start; it != file.end(); ++it)
+    {
+      if (inside)
+      {
+        if (*it == ']')
+        {
+          // Apply integer range expansion (as needed) to the array, which includes only
+          // digits, commas, and '-'. Exclude the outer square brackets.
+          std::string_view str(file.data() + (start - file.cbegin() + 1), it - start - 1);
+          std::size_t s = 0, pos;
+          output << '[';
+          while ((pos = str.find(',', s)) != std::string::npos)
+          {
+            output << RangeExpand(str.substr(s, pos - s)) << ',';
+            s = pos + 1;
+          }
+          output << RangeExpand(str.substr(s)) << ']';
+          start = it + 1;
+          inside = false;
+        }
+        else if (*it == '[')
+        {
+          output << std::string(start, it);
+          start = it;
+        }
+        else if (range_vals.find(*it) == std::string::npos)
+        {
+          output << std::string(start, it);
+          start = it;
+          inside = false;
+        }
+      }
+      else if (*it == '[')
+      {
+        output << std::string(start, it);
+        start = it;
+        inside = true;
+      }
+    }
+    output << std::string(start, file.end());
+  }
+  return output;
```

Comment:
Beyond the scope of this PR but storing large strings of comma separated integers, only to parse them later on into vectors of `int` is a bit roundabout. I think for numbers less than or equal to 6 digits this comes out ahead (digits + space + ,) compared to the 8 bytes of an integer. I think this is still the best approach though, as otherwise this would get way more complicated, so any perf hit here is moot really.

Reply:
The JSON spec requires the arrays to be comma separated lists of integers, so there isn't really a way around this before passing to the JSON parser.